### ISSUE TITLE
Getting nan as train loss when running finetune script

### ISSUE THREAD

**Issue by warisqr007** [yanghaha0908/EmoVoice#5]
I want to fine-tune your model (EmoVoice_1.5B) on additional dataset, and wanted to start from your checkpoint on huggingface. To do so on your `examples/tts/scripts/ft_EmoVoice_1.5B.sh` script I pointed the `ckpt_path` to the model I downloaded from HF. For sanity check I am using the EmoVoice-Db dataset itself. For some reason I am getting loss as `nan` and accuracy as `0.0`. Might be something silly that I am missing. Can you please help me fix this?

My config looks like below:
```
[2025-06-04 22:34:59][root][INFO] - train_config: {'model_name': 'tts', 'enable_ddp': True, 'enable_deepspeed': False, 'enable_fsdp': False, 'low_cpu_fsdp': False, 'run_validation': True, 'batch_size_training': 4, 'batching_strategy': 'custom', 'context_length': 4096, 'gradient_accumulation_steps': 1, 'num_epochs': 50, 'num_workers_dataloader': 0, 'warmup_steps': 1000, 'total_steps': 300000, 'validation_interval': 2500, 'lr': 5e-06, 'weight_decay': 0.0, 'gamma': 0.85, 'seed': 42, 'use_fp16': True, 'mixed_precision': True, 'val_batch_size': 4, 'use_peft': False, 'peft_config': {'peft_method': 'lora', 'r': 8, 'lora_alpha': 32, 'target_modules': ['q_proj', 'v_proj', 'k_proj', 'o_proj'], 'bias': 'none', 'task_type': 'CAUSAL_LM', 'lora_dropout': 0.05, 'inference_mode': False}, 'output_dir': 'EmoVoice/debug1', 'freeze_layers': False, 'num_freeze_layers': 1, 'quantization': False, 'one_gpu': False, 'save_model': True, 'dist_checkpoint_root_folder': 'PATH/to/save/FSDP/model', 'dist_checkpoint_folder': 'fine-tuned', 'save_optimizer': False, 'use_fast_kernels': False, 'run_test_during_validation': False, 'run_test_during_validation_file': 'test.wav', 'run_test_during_validation_prompt': '<|S2S|>', 'freeze_llm': False, 'freeze_encoder': True, 'train_embed_only': False, 'train_audio_embed_only': False, 'task_type': 'TTS', 'freeze_encoder_projector': False, 'freeze_group_decode_adapter': False, 'modeling_paradigm': 'parallel', 'interleaved_text_token_num': 12, 'interleaved_audio_token_num': 36}
[2025-06-04 22:34:59][root][INFO] - fsdp_config: {'mixed_precision': True, 'use_fp16': True, 'sharding_strategy': 'NO_SHARD', 'checkpoint_type': 'SHARDED_STATE_DICT', 'fsdp_activation_checkpointing': True, 'fsdp_cpu_offload': False, 'pure_bf16': False, 'optimizer': 'AdamW'}
[2025-06-04 22:34:59][root][INFO] - model_config: {'file': 'examples/tts/model/slam_model_tts.py:model_factory', 'llm_name': 'Qwen2.5-1.5b', 'llm_path': 'Qwen/Qwen2.5-1.5B', 'llm_type': 'decoder_only', 'llm_dim': 896, 'encoder_name': None, 'encoder_ds_rate': 2, 'encoder_path': None, 'encoder_dim': 1280, 'encoder_projector': 'linear', 'encoder_projector_ds_rate': 5, 'modal': 'audio', 'normalize': False, 'encoder_type': 'finetune', 'vocab_config': {'text_vocabsize': 151936, 'text_specialtokens': 64, 'audio_vocabsize': 4096, 'audio_specialtokens': 64, 'code_layer': 3, 'padded_text_vocabsize': 152000, 'padded_audio_vocabsize': 4160, 'total_audio_vocabsize': 4160, 'total_vocabsize': 156160, 'eot': 151936, 'pad_t': 151937, 'input_t': 151938, 'answer_t': 151939, 'asr': 151940, 'eoa': 4096, 'pad_a': 4097, 'input_a': 4098, 'answer_a': 4099, 'split': 4100}, 'codec_decode': False, 'codec_decoder_type': 'SNAC', 'codec_decoder_path': None, 'encoder_path_hf': None, 'code_type': 'SNAC', 'group_decode': True, 'group_decode_adapter_type': 'linear', 'whisper_decode': False, 'cosyvoice_version': 1, 'random_init': False, 'use_text_stream': True, 'save_audio_token': False, 'phn_tokenizer': None}
[2025-06-04 22:35:40][root][INFO] - dataset_config: {'dataset': 'speech_dataset_tts', 'file': 'examples/tts/speech_dataset_tts.py:get_speech_dataset', 'train_data_path': 'EmoVoice-DB/train.jsonl', 'val_data_path': 'EmoVoice-DB/val.jsonl', 'train_split': 'train', 'test_split': 'validation', 'prompt': None, 'data_path': None, 'max_words': None, 'max_mel': None, 'fix_length_audio': -1, 'inference_mode': False, 'normalize': False, 'seed': 42, 'split_size': 0.01, 'vocab_config': {'text_vocabsize': 151936, 'text_specialtokens': 64, 'audio_vocabsize': 4096, 'audio_specialtokens': 64, 'code_layer': 3, 'padded_text_vocabsize': 152000, 'padded_audio_vocabsize': 4160, 'total_audio_vocabsize': 4160, 'total_vocabsize': 156160, 'eot': 151936, 'pad_t': 151937, 'input_t': 151938, 'answer_t': 151939, 'asr': 151940, 'eoa': 4096, 'pad_a': 4097, 'input_a': 4098, 'answer_a': 4099, 'split': 4100}, 'load_from_cache_file': False, 'task_type': 'TTS', 'upsampling_factor': 1, 'upsample_method': 'repeat', 'num_latency_tokens': 0, 'do_layershift': False, 'use_emo': True, 'use_text_stream': False, 'modeling_paradigm': 'parallel', 'interleaved_text_token_num': 12, 'interleaved_audio_token_num': 36}
```

The only thing I can think of is the Qwen LLM params which I set as follows (since code uses transformers library): 
```
llm_path="Qwen/Qwen2.5-1.5B"
llm_name=Qwen2.5-1.5b
llm_dim=896                         # 896 1536 3584 8192  -> 0.5B 1.5B 3B 7B
```
Do I need to set it differently?

Also, training log for your reference:
```
Training Epoch: 1/50, step 18/3947 completed (loss: nan, audio_acc: 0.0000, text_acc: 0.0000):   0%|▎                                                               | 19/3947 [00:12<44:14,  1.48it/s]
```

**Comment by yanghaha0908** [comment]
Based on the information above, nothing seems out of place. What might be problematic, however, is that during the emotion fine‑tuning phase I used the emovoice‑db along with 200 hours of LAION’s Got Talent data—but I don’t think that has a major impact. Could you look into exactly where, or which variable, is causing the loss to turn into NaN?

**Comment by warisqr007** [comment]
Size of dataset shouldn't matter, like you said.

For me the output of the LLM [here](https://github.com/yanghaha0908/EmoVoice/blob/main/examples/tts/model/slam_model_tts.py#L201) were all `nan`. The input to `self.llm` seems fine. I am still investigating further, would be helpful to get some insights.

Interestingly I tried the script  [here](https://gist.github.com/warisqr007/98109e99b7068b6b257ea9058e605a33) for inference, just so I can do my own inference for each utterance, I wasn't facing any issue. That led me to think might have to do with the setup.

**Comment by yanghaha0908** [comment]
I think you could try lowering the learning rate and also check whether the data precision is correct.
Could you check whether you encounter the same issue if you don't load the checkpoint I trained?

**Comment by warisqr007** [comment]
Yes I had to mark `use_fp16=false`. Then it works. Thanks for the help. Still not sure why with `use_fp16=true` its having issues. I tried without loading your checkpoint as well.

**Comment by yanghaha0908** [comment]
Strange indeed, I had no problem pre-training and fine-tuning with `use_fp16=true`before.
Did lowering the learning rate help?
